iceoryx_posh/source/roudi/process_manager.cpp

```diff
@@ -472,321 +406,269 @@ bool ProcessManager::removeProcess(const std::lock_guard<std::mutex>& lockGuard
 
-        runtime::IpcMessage sendBuffer;
-        sendBuffer << runtime::IpcMessageTypeToString(runtime::IpcMessageType::CREATE_APPLICATION_ACK)
-                   << std::to_string(offset) << std::to_string(m_mgmtSegmentId);
-        process->sendViaIpcChannel(sendBuffer);
+            runtime::IpcMessage sendBuffer;
+            sendBuffer << runtime::IpcMessageTypeToString(runtime::IpcMessageType::CREATE_APPLICATION_ACK)
+                       << std::to_string(offset) << std::to_string(m_mgmtSegmentId);
+            process.sendViaIpcChannel(sendBuffer);
 
-        LogDebug() << "Created new ApplicationPort for application " << name;
-    }
-    else
-    {
-        LogWarn() << "Unknown application " << name << " requested an ApplicationPort." << name;
-    }
+            LogDebug() << "Created new ApplicationPort for application " << name;
+        },
+        [&]() { LogWarn() << "Unknown application " << name << " requested an ApplicationPort." << name; });
 }
 
 void ProcessManager::addNodeForProcess(const ProcessName_t& processName, const NodeName_t& nodeName) noexcept
 {
-    std::lock_guard<std::mutex> g(m_mutex);
-
-    RouDiProcess* process = getProcessFromList(processName);
-    if (nullptr != process)
-    {
-        m_portManager.acquireNodeData(processName, nodeName)
-            .and_then([&](auto nodeData) {
-                auto offset = RelativePointer::getOffset(m_mgmtSegmentId, nodeData);
-
-                runtime::IpcMessage sendBuffer;
-                sendBuffer << runtime::IpcMessageTypeToString(runtime::IpcMessageType::CREATE_NODE_ACK)
-                           << std::to_string(offset) << std::to_string(m_mgmtSegmentId);
-
-                process->sendViaIpcChannel(sendBuffer);
-                m_processIntrospection->addNode(ProcessName_t(cxx::TruncateToCapacity, processName.c_str()),
-                                                NodeName_t(cxx::TruncateToCapacity, nodeName.c_str()));
-                LogDebug() << "Created new node " << nodeName << " for process " << processName;
-            })
-            .or_else([&](PortPoolError error) {
-                runtime::IpcMessage sendBuffer;
-                sendBuffer << runtime::IpcMessageTypeToString(runtime::IpcMessageType::ERROR);
-                if (error == PortPoolError::NODE_DATA_LIST_FULL)
-                {
-                    sendBuffer << runtime::IpcMessageErrorTypeToString(
-                        runtime::IpcMessageErrorType::NODE_DATA_LIST_FULL);
-                }
-                process->sendViaIpcChannel(sendBuffer);
-
-                LogDebug() << "Could not create new node for process " << processName;
-            });
-    }
-    else
-    {
-        LogWarn() << "Unknown process " << processName << " requested a node.";
-    }
+    searchForProcessAndThen(
+        processName,
+        [&](Process& process) {
+            m_portManager.acquireNodeData(processName, nodeName)
+                .and_then([&](auto nodeData) {
```

Comment:
I thought we had a rule to not call a lambda within a lambda. Not sure though.

Reply:
@elBoberido I think we can never enforce such a rule. Just think of a lambda which is calling a function, which is calling a function, which is calling a function, which is calling a function, which is calling a function, which is calling a function,
which is calling a lambda.

Reply:
@elfenpiff well, you can if you forbid calling functions in lambdas :laughing: ... it would be impractical, though

Reply:
Thanks for raising this point @elBoberido IIRC there is a MISRA rule. Couldn't find it by a quick search. However, the AUTOSAR guideline states on [P93](https://www.autosar.org/fileadmin/user_upload/standards/adaptive/18-10/AUTOSAR_RS_CPP14Guidelines.pdf), that nested lambdas reduce the readablity of the code. It is tagged _Advisory_ and not _Required_. In general I agree with this statement, on the other side with our error handling conception `cxx::expected` I'm already aware of a dozen of occurences. I'd also vote not to forbid nesting lambdas. Longing for Axivion CI scan...